GPORCA: Fix cardinality estimation for plan nodes with predicates of the form: date_attribute = constant_value

maksm90 opened the pull request:
### Description
Details and scenario to reproduce case is described in issue https://github.com/greenplum-db/gpdb/issues/11169

### Patch description
The current patch replaces mapping for date values from DOUBLE number of microseconds to LINT number of days that is in its turns natural representation of date value in PostgreSQL. Such modification doesn't allow to compare dates with other time related typed values in the context of statistics estimation routines. But GPORCA in its current version (as opposed to PostgreSQL optimizer) doesn't consider such comparisons and assigns to such predicates dummy frequency value and hence dummy cardinality to the plan node that incorporate these predicates.

### Notes for reviewers

- All changes in *.h and *.cpp are about mapping logic
- `src/backend/gporca/server/CMakeLists.txt` and `src/backend/gporca/data/dxl/minidump/EqualityPredicateOverDate.mdp` add new test case for equality predicate where date type attribute participates
- in `src/backend/gporca/data/dxl/minidump/Select-Over-PartTbl.mdp` the cardinality of selection with predicate `date_attribute < constant` is repaired
- in all others *.mdp files including that was mentioned at previous point `DoubleValue` xml attributes for date type column statistics are rewritten to `LintValue` one with appropriate recalculation of attribute value

### Affected GPDB versions
master and below

zellerh commented on src/backend/gporca/data/dxl/minidump/EqualityPredicateOverDate.mdp:
```diff
@@ -0,0 +1,318 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<dxl:DXLMessage xmlns:dxl="http://greenplum.com/dxl/2010/12/">
```
Could you add a comment here with the test case, so this MDP can be regenerated when needed and so people can tell what the intent of the test was? See file AddRedistributeBeforeInsert-1.mdp for an example what XML format to use and what info to put into the comment.

maksm90 replied:
Thanks for the note, added but I'm not sure that I have correctly given query that fills `test` table. At least, locally I cannot get the similar MCV that is published inside MDP file. I'll try to play with filling query to get statistics nearer to that one in MDP.

zellerh reviewed (commented):
Thank you for your PR, @maksm90! Your change looks good to me and I'm happy with the change to map date values to LINT values for statistics computations.

I tried a few test cases and couldn't see a difference in cardinality estimations between your code and master. Here is what I tried:

```
drop table if exists tdate1, tts1;

create table tdate1(a int, b date);
create table tts1(a int, b timestamp);

insert into tdate1 select i, date '1990-01-01' + i * interval '1' day from generate_series(0, 7999) i;
insert into tts1 select a, b::timestamp from tdate1;

analyze tdate1;
analyze tts1;

---------------- date op const conditions ----------------------------------------
-- expect 8000 rows
explain select * from tdate1;
-- 365 rows
explain select * from tdate1 where b between '1998-01-01' and '1998-12-31';
-- 3 rows
explain select * from tdate1 where b in ('1998-01-01', '1998-02-01', '1998-03-01');
explain select * from tdate1 where b = '1998-01-01' or b = '1998-02-01' or b = '1998-03-01';
-- 1 row
explain select * from tdate1 where b = '1998-01-01';

---------------- join conditions with date type ----------------------------------
-- 8000 rows
explain
select *
from tdate1 x join tdate1 y on x.b=y.b;
-- 365 rows
explain
select *
from tdate1 x join tdate1 y on x.b=y.b
where x.b < '1991-01-01';
-- 3 rows
explain
select *
from tdate1 x join tdate1 y on x.b=y.b
where x.b in ('1998-01-01', '1998-02-01', '1998-03-01');
-- 1 row
explain
select *
from tdate1 x join tdate1 y on x.b=y.b
where x.b = '2004-01-10';

---------- condition involving different datatypes, date and timestamp -----------
-- should be 8000, actual estimate is 25,600,000
explain
select *
from tdate1 join tts1 on tdate1.b = tts1.b;
-- should be ~0, actual value is ~3.2 million
explain
select *
from tdate1 join tts1 on tdate1.b = tts1.b
where tdate1.b > '2000-01-01' and tts1.b < '1995-01-01 00:00:00';
-- should be 3, actual value is 12,800
explain
select *
from tdate1 join tts1 on tdate1.b = tts1.b
where tdate1.b in ('1998-01-01', '1998-02-01', '1998-03-01');

------- condition involving different datatypes, force date comparison -----------
-- 8000
explain
select *
from tdate1 join tts1 on tdate1.b = tts1.b::date;
-- should be ~0, actual value is 1827
explain
select *
from tdate1 join tts1 on tdate1.b = tts1.b::date
where tdate1.b > '2000-01-01' and tts1.b < '1995-01-01 00:00:00';
-- should be 3, actual value is 10,667
explain
select *
from tdate1 join tts1 on tdate1.b = tts1.b::date
where tdate1.b in ('1998-01-01', '1998-02-01', '1998-03-01');
```

Did I make a mistake or did I not try the right types of queries? The good news is of course that your code didn't make any of these tests worse, so this is not a real problem. I was also worried that it might cause trouble when we compare dates with timestamps, but from what I can tell that's not an issue, either.

So again, looks good, just have one small comment on the added MDP and one question.

zellerh commented:
Did a bit more thinking and hope I found the test case: A narrow interval of values when the table contains duplicate date values:

```
drop table if exists tdate2;
create table tdate2(a int, b date);
-- fill the table for 3 years (1990, 1991, 1992) with 1000 rows per day
insert into tdate2 select i%1100, date '1990-01-01' + i%1100 * interval '1' day from generate_series(1, 1100000) i;
analyze tdate2;

-- expect 1.1 M rows
explain select * from tdate2;
-- expect 365 K rows
explain select * from tdate2 where b between '1991-01-01' and '1991-12-31';
-- expect 3000 rows, get the expected estimate in both master and PR
explain select * from tdate2 where b in ('1992-01-01', '1992-02-01', '1992-03-01');
-- expect 3000 rows, get 2 rows in master, 2790 with PR
explain select * from tdate2 where b = '1992-01-01' or b = '1992-02-01' or b = '1992-03-01';
-- expect 1000 rows, get 1 row in master, 903 with PR
explain select * from tdate2 where b = '1992-01-01';
-- check histogram statistics - looks good as far as I can tell
select * from pg_statistic where starelid = 'tdate2'::regclass and staattnum = 2;
```

That is a very nice improvement! The changed estimates are a bit off (seeing both more and less than the estimated value). I see a similar estimation error with the integer column, so this does not look like a problem with your code.

zellerh commented:
@maksm90, could you look at the feedback and provide your test case in the MDP file, if possible? If we don't hear from you we will make the necessary changes to merge the PR.

maksm90 commented:
@zellerh Hi, sorry for so long delay in current work. I have rebased PR branch and added explanatory comment to my test case in MDP file.
As for your provided cases with not improved cardinalities, I need more time to investigate this cases. I'll clearly answer later. This requires from me a deep investigation of logic of calculating cardinality for different types of predicates and joins and hence will consumes some time.

zellerh reviewed (approved):
Thanks for adding the test case. The changes look good to me.

zellerh commented:
> As for your provided cases with not improved cardinalities, I need more time to investigate this cases. I'll clearly answer later. This requires from me a deep investigation of logic of calculating cardinality for different types of predicates and joins and hence will consumes some time.

The reasons for this are probably in other areas of the code, so if it's ok with you we can worry about these test cases another time. The change you have in this PR makes a good set of features to go in. Unless there are objections I'll work on merging this PR in the coming days.

maksm90 commented:
@zellerh thanks for merging this work. Soon I'll present backport to 6X.

zellerh commented:
Thanks, @maksm90. A backport PR for 6X would be great!
